### PR TITLE
[FGS]: function graph resource

### PR DESCRIPTION
**Pull request by artem-lifshits**
## Summary of the Pull Request
Implementation of `opentelekomcloud_fgs_function_v2` base resource.

## PR Checklist

* [ ] Refers to: #xxx
* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.
* [x] Release notes added.

## Acceptance Steps Performed

```
=== RUN   TestAccFgsV2Function_basic
=== PAUSE TestAccFgsV2Function_basic
=== CONT  TestAccFgsV2Function_basic
--- PASS: TestAccFgsV2Function_basic (81.45s)
PASS

Process finished with the exit code 0


=== RUN   TestAccFgsV2Function_text
=== PAUSE TestAccFgsV2Function_text
=== CONT  TestAccFgsV2Function_text
--- PASS: TestAccFgsV2Function_text (29.65s)
PASS

Process finished with the exit code 0

=== RUN   TestAccFgsV2Function_logConfig
=== PAUSE TestAccFgsV2Function_logConfig
=== CONT  TestAccFgsV2Function_logConfig
--- PASS: TestAccFgsV2Function_logConfig (46.93s)
PASS

Process finished with the exit code 0

=== RUN   TestAccFgsV2Function_strategy
=== PAUSE TestAccFgsV2Function_strategy
=== CONT  TestAccFgsV2Function_strategy
--- PASS: TestAccFgsV2Function_strategy (85.98s)
PASS

Process finished with the exit code 0

=== RUN   TestAccFgsV2Function_versions
=== PAUSE TestAccFgsV2Function_versions
=== CONT  TestAccFgsV2Function_versions
--- PASS: TestAccFgsV2Function_versions (67.60s)
PASS

Process finished with the exit code 0

=== RUN   TestAccFgsV2Function_reservedInstance_version
=== PAUSE TestAccFgsV2Function_reservedInstance_version
=== CONT  TestAccFgsV2Function_reservedInstance_version
--- PASS: TestAccFgsV2Function_reservedInstance_version (49.36s)
PASS

Process finished with the exit code 0

=== RUN   TestAccFgsV2Function_concurrencyNum
=== PAUSE TestAccFgsV2Function_concurrencyNum
=== CONT  TestAccFgsV2Function_concurrencyNum
--- PASS: TestAccFgsV2Function_concurrencyNum (47.30s)
PASS

Process finished with the exit code 0


```

### CODE REVIEW

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go [hunk 1]
```diff
@@ -0,0 +1,1193 @@
...
+										Description: "The name of the version alias.",
+									},
+									"description": {
+										Type:        schema.TypeString,
+										Optional:    true,
+										Description: "The description of the version alias.",
+									},
+								},
+							},
+							Description: "The aliases management for specified version.",
```
Comment (anton-sidelnikov) [inline opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go]: remove description on all elements

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go [hunk 2]
```diff
@@ -0,0 +1,1193 @@
...
+		mountConfig.MountUser = mountUser
+	}
+	return &mountConfig
+}
+
+/*
+ * Parse urn according from fun_urn.
+ * If the separator is not ":" then return to the original value.
+ */
+func resourceFgsFunctionUrn(urn string) string {
```
Comment (anton-sidelnikov) [inline opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go]: move to common

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go [hunk 3]
```diff
@@ -0,0 +1,1193 @@
...
+		if err = updateReservedInstanceConfig(fgsClient, d); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return resourceFgsFunctionV2Read(ctx, d, meta)
+}
+
+func resourceFgsFunctionV2Delete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*cfg.Config)
```
Comment (anton-sidelnikov) [inline opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go:1022]: please use context for all clients

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go [hunk 4]
```diff
@@ -0,0 +1,1193 @@
...
+		}
+	}
+
+	if d.HasChanges("reserved_instances") {
+		if err = updateReservedInstanceConfig(fgsClient, d); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return resourceFgsFunctionV2Read(ctx, d, meta)
```
Comment (anton-sidelnikov) [inline opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go]: from context

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go [hunk 5]
```diff
@@ -0,0 +1,1193 @@
...
+
+	return nil
+}
+
+func resourceFgsFunctionV2Update(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*cfg.Config)
+	fgsClient, err := config.FuncGraphV2Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating FunctionGraph v2 client: %s", err)
+	}
```
Comment (anton-sidelnikov) [inline opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go]: from context

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go [hunk 6]
```diff
@@ -0,0 +1,1193 @@
...
+func getConcurrencyNum(concurrencyNum *int) int {
+	return *concurrencyNum
+}
+
+func resourceFgsFunctionV2Read(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*cfg.Config)
+	fgsClient, err := config.FuncGraphV2Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating FunctionGraph client: %s", err)
+	}
```
Comment (anton-sidelnikov) [inline opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go]: from context

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go [hunk 7]
```diff
@@ -0,0 +1,1193 @@
...
+	return result, nil
+}
+
+func resourceFgsFunctionV2Create(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*cfg.Config)
+	fgsClient, err := config.FuncGraphV2Client(config.GetRegion(d))
+	if err != nil {
+		return diag.Errorf("error creating FunctionGraph v2 client: %s", err)
+	}
```
Comment (anton-sidelnikov) [inline opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go]: from context

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go [hunk 8]
```diff
@@ -0,0 +1,1193 @@
...
+		return diag.Errorf("error creating function versions: %s", err)
+	}
+
+	if d.HasChanges("reserved_instances") {
+		if err = updateReservedInstanceConfig(fgsClient, d); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return resourceFgsFunctionV2Read(ctx, d, meta)
```
Comment (anton-sidelnikov) [inline opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go]: from context

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go (hunk 1 above)
Comment (artem-lifshits) [reply opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go]: fixed

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go (hunk 2 above)
Comment (artem-lifshits) [reply opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go]: fixed

File: opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go (hunk 3 above)
Comment (artem-lifshits) [reply opentelekomcloud/services/fgs/resource_opentelekomcloud_fgs_function_v2.go:1022]: fixed

**Review by anton-sidelnikov — APPROVED** [review]

**Pull request merged**